### PR TITLE
Proxy, ProxyFactory and UserWalletFactory contracts.

### PR DESCRIPTION
**Pull request by pgev**
- Proxy, ProxyFactory and UserWalletFactory implementation.
- TokenHolder authorizes session keys during its initialization.

### CODE REVIEW

File: contracts/UserWalletFactory.sol [hunk 1]
```diff
@@ -0,0 +1,124 @@
+pragma solidity ^0.5.0;
+
+// Copyright 2018 OpenST Ltd.
```
Comment (benjaminbollen) [inline contracts/UserWalletFactory.sol]: `Copyright 2019 OpenST Ltd.`

File: contracts/UserWalletFactory.sol [hunk 2]
```diff
@@ -0,0 +1,124 @@
...
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import "./ProxyFactory.sol";
+
+/**
+ * @title Allows to create a new gnosis safe proxy and executes a
+ *        message call to the newly proxy. Afterwards, in the same
```
Comment (benjaminbollen) [inline contracts/UserWalletFactory.sol]: "newly [created] proxy"

File: contracts/UserWalletFactory.sol [hunk 3]
```diff
@@ -0,0 +1,124 @@
...
+contract UserWalletFactory {
+
+    /* Events */
+
+    /**
+     * @notice The event is emitted the from UserWalletFactory::createWallet
+     *         function on success.
+     *
+     * @param _gnosisSafeProxy A newly created gnosis safe's proxy.
+     * @param _tokenHolderProxy A newly crated token holder's proxy.
```
Comment (benjaminbollen) [inline contracts/UserWalletFactory.sol]: typo

File: contracts/UserWalletFactory.sol [hunk 4]
```diff
@@ -0,0 +1,124 @@
...
+        keccak256(
+            "setup(address,address,address,address[],uint256[],uint256[])"
+        )
+    );
+
+
+    /* External Functions  */
+
+    /**
+    * @notice Create a new gnosis safe proxy and executes a
```
Comment (benjaminbollen) [inline contracts/UserWalletFactory.sol]: similar typos as L20

File: contracts/test_doubles/unit_tests/MasterCopySpy.sol [hunk 5]
```diff
@@ -0,0 +1,91 @@
...
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+import "../../SafeMath.sol";
+
+contract MasterCopySpy {
```
Comment (benjaminbollen) [inline contracts/test_doubles/unit_tests/MasterCopySpy.sol]: can be simplified to `MasterCopyFooSpy`; not urgent

File: test/proxy/fallback.js [hunk 6]
```diff
@@ -0,0 +1,218 @@
...
+        const events = Event.decodeTransactionResponse(
+            transactionResponse,
+        );
+
+        assert.strictEqual(
+            events.length,
+            1,
+        );
+
+        Event.assertEqual(events[0], {
```
Comment (benjaminbollen) [inline test/proxy/fallback.js]: would be good to additionally assert in which contract the event is emitted.

File: contracts/ProxyFactory.sol [hunk 7]
```diff
@@ -0,0 +1,67 @@
+pragma solidity ^0.5.0;
+
+// Copyright 2018 OpenST Ltd.
```
[the remark this answers is not in the document]
Comment (pgev) [reply contracts/ProxyFactory.sol]: Done.

File: contracts/TokenHolder.sol [hunk 8]
```diff
@@ -102,6 +102,16 @@ contract TokenHolder {
 
     /* Storage */
 
+    /**
+     * @dev THIS STORAGE VARIABLE *MUST* BE ALWAYS THE FIRST STORAGE
+     *      ELEMENT FOR THIS CONTRACT.
+     *
+     *      This contract is proxied and acts as a master copy for a proxy
```
[the remark this answers is not in the document]
Comment (pgev) [reply contracts/TokenHolder.sol]: Thx for better text. Done.

File: contracts/TokenHolder.sol [hunk 9]
```diff
@@ -147,25 +157,50 @@ contract TokenHolder {
...
-    /* Special Functions */
+    /* External Functions */
 
     /**
-     * @dev Constructor requires:
-     *          - EIP20 token address is not null.
+     * @notice Setups an already deployed contract.
+     *
+     * @dev The function acts as a "constructor" to the contact and initializes
+     *      its internals. As this contract serves as a master copy for
```
[the remark this answers is not in the document]
Comment (pgev) [reply contracts/TokenHolder.sol]: Done.

File: contracts/TokenHolder.sol [hunk 10]
```diff
@@ -147,25 +157,50 @@ contract TokenHolder {
...
+        address[] calldata _sessionKeys,
+        uint256[] calldata _sessionKeysSpendingLimits,
+        uint256[] calldata _sessionKeysExpirationHeights
     )
-        public
+        external
     {
+        // Assures that function can be called only once.
+        require(
+            address(token) == address(0) &&
```
[the remark this answers is not in the document]
Comment (pgev) [reply contracts/TokenHolder.sol]: Actually & is a stronger check. If it's confusing though I can only check `owner`.

File: contracts/TokenHolder.sol [hunk 11]
```diff
@@ -210,19 +260,7 @@ contract TokenHolder {
         keyIsNotNull(_sessionKey)
         keyDoesNotExist(_sessionKey)
```
[the remark this answers is not in the document]
Comment (pgev) [reply contracts/TokenHolder.sol]: Agree. Removed.

File: contracts/UserWalletFactory.sol (hunk 1 above)
Comment (pgev) [reply contracts/UserWalletFactory.sol]: Done.

File: contracts/UserWalletFactory.sol (hunk 2 above)
Comment (pgev) [reply contracts/UserWalletFactory.sol]: Done.

File: contracts/UserWalletFactory.sol (hunk 3 above)
Comment (pgev) [reply contracts/UserWalletFactory.sol]: Done.

File: contracts/UserWalletFactory.sol (hunk 4 above)
Comment (pgev) [reply contracts/UserWalletFactory.sol]: Done.

File: test/proxy/fallback.js (hunk 6 above)
Comment (pgev) [reply test/proxy/fallback.js]: Agree. I am planing to do this for all other event testing. There is a ticket in my list. Would suggest to do this after `proxy`, `credit` and `delayed recovery module` functionalities are merged.

File: contracts/test_doubles/unit_tests/MasterCopySpy.sol (hunk 5 above)
Comment (pgev) [reply contracts/test_doubles/unit_tests/MasterCopySpy.sol]: Will improve later. Currently, have added documentation to make the contract's intend clearer.

File: contracts/TokenHolder.sol (hunk 10 above)
[the remark this answers is not in the document]
Comment (benjaminbollen) [reply contracts/TokenHolder.sol]: 🤦‍♂️ thanks

**Pull request merged**